### PR TITLE
Add findInPath method to locate all matching executables in the system path

### PR DESCRIPTION
**Pull request by smorimoto**
David is the original author. This PR is mainly for #610.

### CODE REVIEW

**Comment by smorimoto** [comment]
Could you check this? @ericsciple @bryanmacfarlane

**Comment by smorimoto** [comment]
CC: @konradpabjan @damccorm

**Comment by smorimoto** [comment]
We won't be able to work on fixing a lot of cache-related issues until we've done this.

**Comment by smorimoto** [comment]
@joshmgross Hi Josh. Sorry to keep pinging you all, but this is one of the steps needed to speed up actions/cache and avoid some bugs. Could you check it?

**Comment by smorimoto** [comment]
ping

**Comment by smorimoto** [comment]
@dhadka as well as #698

**Comment by Be-ing** [comment]
ping

**Comment by smorimoto** [comment]
@joshmgross @dhadka When using cache action on Windows with the Cygwin environment, we have to merge the PR that requires this change, and the delay in reviewing this is literally blocking all those tasks.

**Comment by Be-ing** [comment]
ping! Please prioritize whatever it takes to use zstd reliably on Windows. Decompressing a 1 GB cache on Windows takes 5 minutes per job!

**Comment by smorimoto** [comment]
@dhadka Can you assign a person about this? Other team members seem to have no interest in working on this topic.

**Review by dhadka — APPROVED** [review]
This looks fine to me.

**Comment by dhadka** [comment]
@actions/actions-runtime Can we please get a review on this PR?

**Review by konradpabjan — COMMENTED** [review]
Changes look good.

If you could add some tests though that would be great! Good to have for future maintenance. We have some existing ones here: https://github.com/actions/toolkit/blob/main/packages/io/__tests__/io.test.ts

**Comment by smorimoto** [comment]
There are a lot of irrelevant changes after I changed what I was concerned, so I will do them in another PR. Anyway, this PR is ready to merge.

**Comment by dhadka** [comment]
Thanks!  @konradpabjan can you please give this another look and a ✔️ for `actions-service` if OK?  `findInPath` is used by `io.which` so it should have good test coverage in addition to the test @smorimoto added.

**Comment by smorimoto** [comment]
Actually, the coverage isn't as bad as we thought.

```
------------|---------|----------|---------|---------|----------------------------------------------------------------------------------
File        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
------------|---------|----------|---------|---------|----------------------------------------------------------------------------------
All files   |   74.02 |    65.96 |   94.12 |   75.63 |
 io-util.ts |    64.2 |    53.85 |    87.5 |    64.1 | 29,50,54,125,133-135,147,149-151,153,155,161,162,164-170,175,180,182,183,196,199
 io.ts      |   80.49 |    73.03 |     100 |   83.19 | 54,69,98,117-119,121,126,130,131,135,182,204,215,221-223,323,327,328
------------|---------|----------|---------|---------|----------------------------------------------------------------------------------
```

**Comment by smorimoto** [comment]
@dhadka As for actions-service, the review request is made because of the unrelated commit I excluded, so we can ignore it.

**Review by konradpabjan — APPROVED** [review]
Looks good to me! Thanks for adding some tests. Going to ask one more person for a quick  👀  though before merging this in and publishing

**Comment by smorimoto** [comment]
Thanks again 🙂

File: packages/io/src/io.ts
Comment (thboop) [inline packages/io/src/io.ts:204]: Why are we wrapping this in a try catch? Is there any reason we don't just throw the original exception? This will wrap stuff like `parameter 'tool' is required`

File: packages/io/src/io.ts
Comment (thboop) [inline packages/io/src/io.ts:240]: Nice clean fix here

**Review by thboop — COMMENTED** [review]
Left some minor comments but lets get a new version of this out once they have been addressed!

File: packages/io/src/io.ts
Comment (smorimoto) [reply packages/io/src/io.ts]: I don't know why this was made, since it's based on a cherry-picked commit, but it should be fixed! Good point!

File: packages/io/src/io.ts
Comment (smorimoto) [reply packages/io/src/io.ts]: This has been fixed as well.

**Comment by smorimoto** [comment]
Please let me know if there is anything else I can do to finish.

**Comment by smorimoto** [comment]
ping @dhadka @thboop

**Comment by dhadka** [comment]
Changes LGTM, ready to 🚀 once @thboop gives the 👍

**Comment by smorimoto** [comment]
ping @thboop

**Comment by smorimoto** [comment]
Hi? @thboop

**Comment by thboop** [comment]
Hey @smorimoto ,

I'll take a look at this next week and see if we can get a new release out. We typically have longer turnaround times for new features in open source repositories, as any feature requires further maintenance and support in the long term by our team. I'll be updating the repo with some templates to set better guidance for contributors in the future, I'm sorry for the churn this has caused you.

**Comment by smorimoto** [comment]
Thanks. I can understand the reason itself, but to be honest I'm quite skeptical about the fact that it would be a reasonable reason for hardly do a review for 6 months. And that seems to conflict with the reason why we do versioning in software in general. In any case, I hope we can make a "fix" (not an improvement!) for actions/cache as soon as possible.

**Comment by Be-ing** [comment]
I also think you should reconsider your processes if it takes 6 months to review a rather small pull request.

**Comment by smorimoto** [comment]
I'm glad I'm not the only one who feels that. In fact, the most PRs sent to the Actions repositories are well-known to require 100 pings to move the review process forward. This is a real loss for both GitHub and its users.

**Comment by smorimoto** [comment]
ping?

**Review by thboop — APPROVED** [review]
Changes Look great LGTM

**Comment by dhadka** [comment]
🚀

**Comment by dhadka** [comment]
@konradpabjan Can we please get a new version of this built?

**Comment by thboop** [comment]
Thanks for the contribution @smorimoto and being patient with the delay, working on a new npm release now 🚀 🚀

**Comment by smorimoto** [comment]
That was quick.

**Comment by thboop** [comment]
@actions/io v1.1.0 has been published 🚀 🚀 🚀

**Pull request merged**
